pkg/forwarder/transaction.go

```diff
@@ -111,24 +111,58 @@ type HTTPTransaction struct {
 	ErrorCount int
 
 	createdAt time.Time
+	// retryable indicates whether this transaction can be retried
+	retryable bool
+
+	// attemptHandler will be called with a transaction before the attempting to send the request
+	attemptHandler func(transaction *HTTPTransaction)
+	// completionHandler will be called with a transaction after it has been successfully sent
+	completionHandler func(transaction *HTTPTransaction, statusCode int, body []byte, err error)
 }
 
 // Transaction represents the task to process for a Worker.
 type Transaction interface {
-	Process(ctx context.Context, client *http.Client) error
+	Process(ctx context.Context, client *http.Client) (int, []byte, error)
 	GetCreatedAt() time.Time
 	GetTarget() string
+	IsRetryable() bool
+	OnAttempt()
+	OnComplete(statusCode int, body []byte, err error)
 }
 
+var _ Transaction = &HTTPTransaction{}
+
 // NewHTTPTransaction returns a new HTTPTransaction.
 func NewHTTPTransaction() *HTTPTransaction {
 	return &HTTPTransaction{
 		createdAt:  time.Now(),
 		ErrorCount: 0,
+		retryable:  true,
 		Headers:    make(http.Header),
+		attemptHandler: func(transaction *HTTPTransaction) {
+
```

Comment:
Is  `func(transaction *HTTPTransaction) {}` not accepted by the linter? 

Might be worth extracting these functions to their own global variables so this function is a bit cleaner.

Reply:
I haven't seen any linting errors.  

> Might be worth extracting these functions to their own global variables so this function is a bit cleaner.

Are you suggesting a type alias for the function signature or a global var for the noop implementations?  Or both?

Reply:
global var